File: frontend/src/services/playbooksTasks.ts
import { apiRequest } from '@/services/api'
import { Page } from '@/services/accountWorkspace'

function queryString(params: URLSearchParams) {
  const query = params.toString()
  return query ? `?${query}` : ''
}

export interface PlaybookTemplateActivity {
  id: string
  template_id: string
  title: string
  description?: string | null
  owner_rule: string
  due_offset_days: number
  priority: TaskPriority
  success_criteria: string[]
  skip_allowed: boolean
  requires_evidence: boolean
  sort_order: number
  created_at: string
  updated_at: string
}

export type TaskStatus = 'open' | 'todo' | 'in_progress' | 'done' | 'blocked' | 'cancelled' | 'skipped'
export type TaskPriority = 'low' | 'medium' | 'high' | 'urgent' | 'critical'

export interface PlaybookTemplate {
  id: string
  name: string
  objective: string
  description?: string | null
  signal_types: string[]
  weak_metrics: string[]
  default_owner_rule: string
  due_date_rule: Record<string, unknown>
  success_criteria: string[]
  skip_rules: string[]
  version: number
  is_active: boolean
  created_at: string
  updated_at: string
  activities: PlaybookTemplateActivity[]
  custom_field_values?: Record<string, unknown>
}

export interface TaskEvidence {
  id: string
  task_id: string
  evidence_type: 'note' | 'link' | 'file'
  title?: string | null
  body?: string | null
  url?: string | null
  file_name?: string | null
  file_mime_type?: string | null
  file_size_bytes?: number | null
  created_by_name: string
  created_at: string
}

export interface PlaybookTask {
  id: string
  account_id: string
  engagement_id?: string | null
  playbook_execution_id?: string | null
  template_activity_id?: string | null
  source_type: string
  source_record_id?: string | null
  source_metric?: string | null
  title: string
  description?: string | null
  owner_id?: string | null
  owner_name: string
  due_at: string
  status: TaskStatus
  priority: TaskPriority
  notes?: string | null
  outcome?: string | null
  success_criteria: string[]
  requires_evidence: boolean
  skipped_reason?: string | null
  completed_at?: string | null
  evidence: TaskEvidence[]
  custom_field_values?: Record<string, unknown>
  created_at: string
  updated_at: string
}

export interface PlaybookExecution {
  id: string
  template_id?: string | null
  template_name_snapshot: string
  template_version_snapshot: number
  account_id: string
  engagement_id?: string | null
  source_signal_id?: string | null
  source_signal_type?: string | null
  source_metric?: string | null
  status: string
  tasks: PlaybookTask[]
  created_at: string
}

export interface RecommendedPlaybook {
  template: PlaybookTemplate
  rationale: string
  match_score: number
  matched_signal_types: string[]
  matched_metrics: string[]
}

export interface CalendarItem {
  id: string
  kind: 'governance' | 'governance_action' | 'task' | 'sow_expiry' | 'renewal_date' | 'notice_deadline' | string
  title: string
  detail: string
  account_id?: string | null
  account_name: string
  engagement_id?: string | null
  owner_id?: string | null
  owner_name?: string | null
  date: string
  status: string
  priority?: string | null
  source_route?: string | null
  source_record_id?: string | null
  source_record_type?: string | null
}

export interface PlaybookTemplatePayload {
  name: string
  objective: string
  description?: string
  signal_types?: string[]
  weak_metrics?: string[]
  default_owner_rule?: string
  due_date_rule?: Record<string, unknown>
  success_criteria: string[]
  skip_rules?: string[]
  activities: Array<{
    title: string
    description?: string
    owner_rule?: string
    due_offset_days?: number
    priority?: TaskPriority
    success_criteria?: string[]
    skip_allowed?: boolean
    requires_evidence?: boolean
    sort_order?: number
  }>
  is_active?: boolean
  custom_field_values?: Record<string, unknown>
}

export async function listPlaybookTemplates(token: string, params = new URLSearchParams()) {
  return apiRequest<Page<PlaybookTemplate>>(`/api/admin/playbook-templates${queryString(params)}`, { token })
}

export async function createPlaybookTemplate(token: string, payload: PlaybookTemplatePayload) {
  return apiRequest<PlaybookTemplate>('/api/admin/playbook-templates', { method: 'POST', token, body: JSON.stringify(payload) })
}

export async function updatePlaybookTemplate(token: string, templateId: string, payload: Partial<PlaybookTemplatePayload>) {
  return apiRequest<PlaybookTemplate>(`/api/admin/playbook-templates/${templateId}`, { method: 'PATCH', token, body: JSON.stringify(payload) })
}

export async function listRecommendedPlaybooks(token: string, signalId: string, params = new URLSearchParams()) {
  return apiRequest<RecommendedPlaybook[]>(`/api/signals/${signalId}/recommended-playbooks${queryString(params)}`, { token })
}

export async function executePlaybook(token: string, templateId: string, payload: { account_id: string; engagement_id?: string; source_signal_id?: string; source_signal_type?: string; source_metric?: string; confirmed: boolean }) {
  return apiRequest<PlaybookExecution>(`/api/playbooks/${templateId}/execute`, { method: 'POST', token, body: JSON.stringify(payload) })
}

export async function listTasks(token: string, params = new URLSearchParams()) {
  return apiRequest<Page<PlaybookTask>>(`/api/tasks${queryString(params)}`, { token })
}

export async function createTask(token: string, payload: Partial<PlaybookTask> & { account_id: string; owner_id: string; title: string; due_at: string; priority: TaskPriority; status: TaskStatus }) {
  return apiRequest<PlaybookTask>('/api/tasks', { method: 'POST', token, body: JSON.stringify(payload) })
}

export async function updateTask(token: string, taskId: string, payload: Partial<Pick<PlaybookTask, 'title' | 'description' | 'owner_id' | 'due_at' | 'status' | 'priority' | 'notes' | 'outcome' | 'skipped_reason' | 'success_criteria' | 'requires_evidence' | 'custom_field_values'>>) {
  return apiRequest<PlaybookTask>(`/api/tasks/${taskId}`, { method: 'PATCH', token, body: JSON.stringify(payload) })
}

export async function addTaskEvidence(token: string, taskId: string, payload: { evidence_type: 'note' | 'link' | 'file'; title?: string; body?: string; url?: string; file?: File | null }) {
  const form = new FormData()
  form.append('evidence_type', payload.evidence_type)
  if (payload.title) form.append('title', payload.title)
  if (payload.body) form.append('body', payload.body)
  if (payload.url) form.append('url', payload.url)
  if (payload.file) form.append('file', payload.file)
  return apiRequest<TaskEvidence>(`/api/tasks/${taskId}/evidence`, { method: 'POST', token, body: form })
}

export async function listCalendarItems(token: string, params = new URLSearchParams()) {
  return apiRequest<Page<CalendarItem>>(`/api/calendar/items${queryString(params)}`, { token })
}

File: backend/app/routers/playbooks_tasks.py
from datetime import datetime
from typing import Annotated, Literal

from fastapi import APIRouter, Depends, Query, Request, UploadFile, status

from app.dependencies import get_current_user, get_playbooks_tasks_service
from app.models import User
from app.schemas import (
    CalendarItemPageRead,
    PlaybookExecutionRead,
    PlaybookExecutionRequest,
    PlaybookTemplateCreateRequest,
    PlaybookTemplatePageRead,
    PlaybookTemplateRead,
    PlaybookTemplateUpdateRequest,
    RecommendedPlaybookRead,
    TaskCreateRequest,
    TaskEvidenceRead,
    TaskPageRead,
    TaskRead,
    TaskUpdateRequest,
)
from app.services.playbooks_tasks import PlaybooksTasksService

Direction = Literal["asc", "desc"]
ActiveState = Literal["all", "active", "inactive"]
TemplateSort = Literal["name", "version", "active_state", "updated_at", "created_at"]
TaskSort = Literal["due_at", "priority", "status", "updated_at", "created_at"]

router = APIRouter(prefix="/api", tags=["Playbooks and Tasks"])


@router.get("/admin/playbook-templates", response_model=PlaybookTemplatePageRead, summary="List playbook templates", description="Paginated playbook template catalog with search, active-state, signal, metric, owner-rule, sort, and RBAC filters.")
def list_templates(
    current_user: Annotated[User, Depends(get_current_user)],
    service: Annotated[PlaybooksTasksService, Depends(get_playbooks_tasks_service)],
    search: str | None = None,
    active_state: ActiveState = "active",
    signal_type: str | None = None,
    weak_metric: str | None = None,
    owner_rule: str | None = None,
    sort: TemplateSort = "updated_at",
    direction: Direction = "desc",
    page: Annotated[int, Query(ge=1)] = 1,
    page_size: Annotated[int, Query(ge=1, le=100)] = 10,
) -> PlaybookTemplatePageRead:
    return service.list_templates(current_user, search=search, active_state=active_state, signal_type=signal_type, weak_metric=weak_metric, owner_rule=owner_rule, sort=sort, direction=direction, page=page, page_size=page_size)


@router.post("/admin/playbook-templates", response_model=PlaybookTemplateRead, status_code=status.HTTP_201_CREATED, summary="Create playbook template", description="Requires playbooks:configure_templates. Creates a configurable playbook template with objectives, activities, owners, due-date rules, success criteria, skip rules, audit, and Field Builder values.")
def create_template(payload: PlaybookTemplateCreateRequest, current_user: Annotated[User, Depends(get_current_user)], service: Annotated[PlaybooksTasksService, Depends(get_playbooks_tasks_service)]) -> PlaybookTemplateRead:
    return service.create_template(payload, current_user)


@router.patch("/admin/playbook-templates/{template_id}", response_model=PlaybookTemplateRead, summary="Update playbook template", description="Requires playbooks:configure_templates. Updates template configuration, versions material changes, and preserves historical execution snapshots.")
def update_template(template_id: str, payload: PlaybookTemplateUpdateRequest, current_user: Annotated[User, Depends(get_current_user)], service: Annotated[PlaybooksTasksService, Depends(get_playbooks_tasks_service)]) -> PlaybookTemplateRead:
    return service.update_template(template_id, payload, current_user)


@router.get("/signals/{signal_id}/recommended-playbooks", response_model=list[RecommendedPlaybookRead], summary="List recommended playbooks", description="Rule-based bridge endpoint that matches active playbook templates to a signal id plus optional signal type/metric context.")
def recommended_playbooks(
    signal_id: str,
    current_user: Annotated[User, Depends(get_current_user)],
    service: Annotated[PlaybooksTasksService, Depends(get_playbooks_tasks_service)],
    signal_type: str | None = None,
    weak_metric: str | None = None,
    account_id: str | None = None,
    engagement_id: str | None = None,
    page: Annotated[int, Query(ge=1)] = 1,
    page_size: Annotated[int, Query(ge=1, le=100)] = 10,
) -> list[RecommendedPlaybookRead]:
    return service.recommended_playbooks(signal_id, current_user, signal_type=signal_type, weak_metric=weak_metric, account_id=account_id, engagement_id=engagement_id, page=page, page_size=page_size)


@router.post("/playbooks/{template_id}/execute", response_model=PlaybookExecutionRead, status_code=status.HTTP_201_CREATED, summary="Execute playbook", description="Requires playbooks:execute and account work access. Executes a confirmed active playbook template and creates account-specific tasks without silently changing health metrics.")
def execute_playbook(template_id: str, payload: PlaybookExecutionRequest, current_user: Annotated[User, Depends(get_current_user)], service: Annotated[PlaybooksTasksService, Depends(get_playbooks_tasks_service)]) -> PlaybookExecutionRead:
    return service.execute_playbook(template_id, payload, current_user)


@router.get("/tasks", response_model=TaskPageRead, summary="List tasks", description="Paginated task list with account, engagement, owner, status, priority, source, due-window, my-items, search, and sort filters.")
def list_tasks(
    current_user: Annotated[User, Depends(get_current_user)],
    service: Annotated[PlaybooksTasksService, Depends(get_playbooks_tasks_service)],
    account_id: str | None = None,
    engagement_id: str | None = None,
    owner_id: str | None = None,
    status_filter: Annotated[str | None, Query(alias="status")] = None,
    priority: str | None = None,
    source_type: str | None = None,
    due_from: datetime | None = None,
    due_to: datetime | None = None,
    search: str | None = None,
    my_items: bool = False,
    sort: TaskSort = "due_at",
    direction: Direction = "asc",
    page: Annotated[int, Query(ge=1)] = 1,
    page_size: Annotated[int, Query(ge=1, le=100)] = 10,
) -> TaskPageRead:
    return service.list_tasks(current_user, account_id=account_id, engagement_id=engagement_id, owner_id=owner_id, status_filter=status_filter, priority=priority, source_type=source_type, due_from=due_from, due_to=due_to, search=search, my_items=my_items, sort=sort, direction=direction, page=page, page_size=page_size)


@router.post("/tasks", response_model=TaskRead, status_code=status.HTTP_201_CREATED, summary="Create task", description="Creates a manual account task/activity with owner, due date, status, priority, notes, success criteria, Field Builder values, audit, and timeline entry.")
def create_task(payload: TaskCreateRequest, current_user: Annotated[User, Depends(get_current_user)], service: Annotated[PlaybooksTasksService, Depends(get_playbooks_tasks_service)]) -> TaskRead:
    return service.create_task(payload, current_user)


@router.patch("/tasks/{task_id}", response_model=TaskRead, summary="Update task", description="Updates task details, status, owner, outcome, skip reason, completion metadata, Field Builder values, audit, and timeline entries.")
def update_task(task_id: str, payload: TaskUpdateRequest, current_user: Annotated[User, Depends(get_current_user)], service: Annotated[PlaybooksTasksService, Depends(get_playbooks_tasks_service)]) -> TaskRead:
    return service.update_task(task_id, payload, current_user)


@router.post("/tasks/{task_id}/evidence", response_model=TaskEvidenceRead, status_code=status.HTTP_201_CREATED, summary="Add task evidence", description="Adds note, link, or file evidence to a task using the existing storage adapter and writes audit/timeline history.")
async def add_task_evidence(
    task_id: str,
    request: Request,
    current_user: Annotated[User, Depends(get_current_user)],
    service: Annotated[PlaybooksTasksService, Depends(get_playbooks_tasks_service)],
) -> TaskEvidenceRead:
    content_type = request.headers.get("content-type", "")
    if "multipart/form-data" in content_type or "application/x-www-form-urlencoded" in content_type:
        form = await request.form()
        evidence_type = str(form.get("evidence_type") or "note")
        title = str(form.get("title")) if form.get("title") else None
        body = str(form.get("body") or form.get("note")) if form.get("body") or form.get("note") else None
        url = str(form.get("url")) if form.get("url") else None
        maybe_file = form.get("file")
        file = maybe_file if hasattr(maybe_file, "filename") else None
    else:
        payload = await request.json()
        evidence_type = str(payload.get("evidence_type") or "note")
        title = payload.get("title")
        body = payload.get("body") or payload.get("note")
        url = payload.get("url")
        file = None
    if evidence_type == "url":
        evidence_type = "link"
    return await service.add_task_evidence(task_id, current_user, evidence_type=evidence_type, title=title, body=body, url=url, file=file)


@router.get("/calendar/items", response_model=CalendarItemPageRead, summary="List unified calendar items", description="Aggregates governance events/action items, playbook tasks, SOW expiry, renewal dates, and notice deadlines with permission-aware filtering.")
def list_calendar_items(
    current_user: Annotated[User, Depends(get_current_user)],
    service: Annotated[PlaybooksTasksService, Depends(get_playbooks_tasks_service)],
    date_from: datetime | None = None,
    date_to: datetime | None = None,
    account_id: str | None = None,
    engagement_id: str | None = None,
    owner_id: str | None = None,
    my_items: bool = False,
    include_governance: bool = True,
    include_tasks: bool = True,
    include_renewals: bool = True,
    search: str | None = None,
    page: Annotated[int, Query(ge=1)] = 1,
    page_size: Annotated[int, Query(ge=1, le=500)] = 100,
) -> CalendarItemPageRead:
    return service.calendar_items(current_user, date_from=date_from, date_to=date_to, account_id=account_id, engagement_id=engagement_id, owner_id=owner_id, my_items=my_items, include_governance=include_governance, include_tasks=include_tasks, include_renewals=include_renewals, search=search, page=page, page_size=page_size)
